### PR TITLE
[node-manager] export NodeGroup metrics

### LINKED ISSUE #5459: [monitoring] export availability metrics for node groups
### Preflight Checklist

- [X] I agree to follow the [Code of Conduct](https://github.com/deckhouse/deckhouse/blob/main/CODE_OF_CONDUCT.md) that this project adheres to.
- [X] I have searched the [issue tracker](https://github.com/deckhouse/deckhouse/issues) for an issue that matches the one I want to file, without success.

### Use case. Why is this important?

```
# kubectl get nodegroups.deckhouse.io worker
NAME     TYPE             READY   NODES   UPTODATE   INSTANCES   DESIRED   MIN   MAX   STANDBY   STATUS   AGE
worker   CloudEphemeral   2       2       2          2           2         2     5                        2y315d
```

Export current status metrics for nodegroups
It's help make alerts of availability nodes on cluster

For example - make metrics for READY / NODES / UPTODATE / INSTANCES / DESIRED / MIN / MAX / STANDBY / AGE

### Proposed Solution

-

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by artbin**
## Description

Add a hook to the `node-manager` module that exports node group metrics.

## Why do we need it, and what problem does it solve?

Close #5459

## Checklist
- [x] The code is covered by unit tests.
- [x] e2e tests passed.
- [ ] Documentation updated according to the changes.
- [x] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: node-manager
type: feature
summary: 'Export availability metrics for node groups. Metrics have the prefix `d8_node_group_`.'
impact_level: default
```

### CODE REVIEW

**Comment by artbin** [comment]
Manual testing
==============

Create the `workers` NodeGroup with `minPerZone: 1` and `maxPerZone: 1`:
```yaml
apiVersion: deckhouse.io/v1
kind: NodeGroup
metadata:
  name: workers
spec:
  nodeType: CloudEphemeral
  cloudInstances:
    minPerZone: 1
    maxPerZone: 1
    classReference:
      kind: OpenStackInstanceClass
      name: workers
```

Change the `workers` NodeGroup with `minPerZone: 5` and `maxPerZone: 10`:
```yaml
apiVersion: deckhouse.io/v1
kind: NodeGroup
metadata:
  name: workers
spec:
  nodeType: CloudEphemeral
  cloudInstances:
    minPerZone: 5
    maxPerZone: 10
    classReference:
      kind: OpenStackInstanceClass
      name: workers
```

An error occurred after updating the NodeGroup specification:
```bash
kubectl get nodegroup workers -o json | jq '.status.conditions[] | select( .type | contains("Error"))'
```
```json
{
"lastTransitionTime": "2024-05-13T04:10:58Z",
"message": "Cloud provider message - error creating the server: Request forbidden: error message: {\"forbidden\": {\"code\": 403, \"message\": \"Quota exceeded for ram: Requested 2048, but already used 24576 of 24576 ram\"}}",
"status": "True",
"type": "Error"
}
```

Change the `workers` NodeGroup with `minPerZone: 2` and `maxPerZone: 4`:
```yaml
apiVersion: deckhouse.io/v1
kind: NodeGroup
metadata:
  name: workers
spec:
  nodeType: CloudEphemeral
  cloudInstances:
    minPerZone: 2
    maxPerZone: 4
    classReference:
      kind: OpenStackInstanceClass
      name: workers
```

<details>
  <summary>Grafana screenshots</summary>
  
Screenshot 2024-05-13 at 12 23 30
Screenshot 2024-05-13 at 12 23 37
Screenshot 2024-05-13 at 12 23 42
Screenshot 2024-05-13 at 12 23 52
Screenshot 2024-05-13 at 12 24 00
Screenshot 2024-05-13 at 12 24 05
Screenshot 2024-05-13 at 12 24 10
Screenshot 2024-05-13 at 12 24 15
Screenshot 2024-05-13 at 12 24 22

</details>

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @artbin. I've started the workflow for label 'e2e/run/openstack'...

File: modules/040-node-manager/hooks/node_group_metrics.go [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+	Name       string
+	Ready      float64
+	Nodes      float64
+	Instances  float64
+	Desired    float64
+	Min        float64
+	Max        float64
+	UpToDate   float64
+	Standby    float64
+	HaseErrors float64
```
Comment (RomanenkoDenys) [inline modules/040-node-manager/hooks/node_group_metrics.go:51]: May be HasErrors ?

**Review by RomanenkoDenys — CHANGES_REQUESTED** [review]
Please add tests

File: modules/040-node-manager/hooks/node_group_metrics.go (hunk 1 above)
Comment (artbin) [reply modules/040-node-manager/hooks/node_group_metrics.go]: fixed

**Review by z9r5 — APPROVED** [review]

File: modules/040-node-manager/docs/README_RU.md [hunk 2]
```diff
@@ -307,3 +307,22 @@ post-install() {
 ## Chaos Monkey
 
 Инструмент (включается у каждой из `NodeGroup` отдельно), позволяющий систематически вызывать случайные прерывания работы узлов. Предназначен для проверки элементов кластера, приложений и инфраструктурных компонентов на реальную работу отказоустойчивости.
+
+## Мониторинг
+
+Группы узлов (ресурс `NodeGroup`) экспортируют метрики доступности, которые обрабатываются Prometheus.
```
Comment (name212) [inline modules/040-node-manager/docs/README_RU.md:313]:
```suggestion
Для групп узлов (ресурс `NodeGroup`) DKP экспортируют метрики доступности группы
```

**Review by name212 — APPROVED** [review]

**Review by RomanenkoDenys — APPROVED** [review]
please see @name212 suggestion

**Review by sprait — APPROVED** [review]

File: modules/040-node-manager/docs/README.md [hunk 3]
```diff
@@ -304,3 +304,24 @@ The script progress can be seen on the node in the bashible service log (`journa
 ## Chaos Monkey
 
 The instrument (you can enable it for each `NodeGroup` individually) for unexpected and random termination of nodes in a systemic manner. Chaos Monkey tests the resilience of cluster elements, applications, and infrastructure components.
+
+## Monitoring 
+
+DKP exports the group availability Prometheus metrics for each `NodeGroup`.
+
+### What information does Prometheus collect, and in what form?
```
Comment (kirkonru) [inline modules/040-node-manager/docs/README.md:312]:
```suggestion
### What information does Prometheus collect and in what form?
```

File: modules/040-node-manager/docs/README.md [hunk 4]
```diff
@@ -304,3 +304,24 @@ The script progress can be seen on the node in the bashible service log (`journa
...
+
+Metrics have the prefix `d8_node_group_`.
```
Comment (kirkonru) [inline modules/040-node-manager/docs/README.md:314]:
```suggestion
Metrics have the `d8_node_group_` prefix.
```

File: modules/040-node-manager/docs/README.md [hunk 5]
```diff
@@ -304,3 +304,24 @@ The script progress can be seen on the node in the bashible service log (`journa
...
+
+All the collected metrics have a label that allow you to identify the node group: `node_group_name`.
+
+Metrics collected for each node group:
```
Comment (kirkonru) [inline modules/040-node-manager/docs/README.md:318]:
```suggestion
The following metrics are collected for each node group:
```

File: modules/040-node-manager/docs/README.md [hunk 6]
```diff
@@ -304,3 +304,24 @@ The script progress can be seen on the node in the bashible service log (`journa
...
+- `d8_node_group_ready` — the number of ready Kubernetes nodes in the group.
+- `d8_node_group_nodes` — the number of Kubernetes nodes (in any state) in the group.
+- `d8_node_group_instances` — the number of instances (in any state) in the group.
+- `d8_node_group_desired` — the number of desired `Machines` in the group.
+- `d8_node_group_min` — the minimal amount of instances in the group.
```
Comment (kirkonru) [inline modules/040-node-manager/docs/README.md:323]:
```suggestion
- `d8_node_group_min` — the minimum number of instances in the group.
```

File: modules/040-node-manager/docs/README.md [hunk 7]
```diff
@@ -304,3 +304,24 @@ The script progress can be seen on the node in the bashible service log (`journa
...
+- `d8_node_group_max` — the maximum amount of instances in the group.
```
Comment (kirkonru) [inline modules/040-node-manager/docs/README.md:324]:
```suggestion
- `d8_node_group_max` — the maximum number of instances in the group.
```

File: modules/040-node-manager/docs/README.md [hunk 8]
```diff
@@ -304,3 +304,24 @@ The script progress can be seen on the node in the bashible service log (`journa
...
+- `d8_node_group_up_to_date` — the number of up-to-date nodes in the group.
+- `d8_node_group_standby` — the number of over provisioned instances (see [standby](cr.html#nodegroup-v1-spec-cloudinstances-standby) parameter) in the group.
```
Comment (kirkonru) [inline modules/040-node-manager/docs/README.md:326]:
```suggestion
- `d8_node_group_standby` — the number of over provisioned instances (see the [standby](cr.html#nodegroup-v1-spec-cloudinstances-standby) parameter) in the group.
```

File: modules/040-node-manager/docs/README.md [hunk 9]
```diff
@@ -304,3 +304,24 @@ The script progress can be seen on the node in the bashible service log (`journa
...
+DKP exports the group availability Prometheus metrics for each `NodeGroup`.
+
+### What information does Prometheus collect, and in what form?
+
+Metrics have the prefix `d8_node_group_`.
+
+All the collected metrics have a label that allow you to identify the node group: `node_group_name`.
+
+Metrics collected for each node group:
+- `d8_node_group_ready` — the number of ready Kubernetes nodes in the group.
```
Comment (kirkonru) [inline modules/040-node-manager/docs/README.md:319]:
```suggestion
- `d8_node_group_ready` — the number of Kubernetes nodes in the group that are `Ready`.
```

File: modules/040-node-manager/docs/README.md [hunk 10]
```diff
@@ -304,3 +304,24 @@ The script progress can be seen on the node in the bashible service log (`journa
...
+- `d8_node_group_nodes` — the number of Kubernetes nodes (in any state) in the group.
+- `d8_node_group_instances` — the number of instances (in any state) in the group.
+- `d8_node_group_desired` — the number of desired `Machines` in the group.
+- `d8_node_group_min` — the minimal amount of instances in the group.
+- `d8_node_group_max` — the maximum amount of instances in the group.
+- `d8_node_group_up_to_date` — the number of up-to-date nodes in the group.
+- `d8_node_group_standby` — the number of over provisioned instances (see [standby](cr.html#nodegroup-v1-spec-cloudinstances-standby) parameter) in the group.
+- `d8_node_group_has_errors` — the boolean value equal to 1 if there are errors in the group.
```
Comment (kirkonru) [inline modules/040-node-manager/docs/README.md:327]:
```suggestion
- `d8_node_group_has_errors` — equals to 1 if there are errors in the group.
```

**Review by kirkonru — APPROVED** [review]

**Review by aleksey-su — APPROVED** [review]

**Pull request merged**